clients/drcachesim/tests/snoop_filter_test.cpp

```diff
@@ -0,0 +1,205 @@
+                /*prefetcher=*/nullptr,
+                /*inclusive=*/true, /*coherence_cache=*/true, i, snoop_filter_);
+        }
+
+        if (!snoop_filter_->init(snooped_caches_, num_cores_)) {
+            std::cerr << "Usage error: failed to initialize snoop filter.\n";
+            return;
+        }
+    }
+
+    void
+    request_and_check_state(const int cache_id, const addr_t addr,
+                            const trace_type_t type, int expected_num_writes,
+                            int expected_num_invalidates, int expected_num_writebacks,
+                            int expected_num_sharers, bool block_is_dirty)
+    {
+        memref_t ref;
+        ref.data.size = 1;
+        ref.data.addr = addr;
+        ref.data.type = type;
+        snooped_caches_[cache_id]->request(ref);
+        snoop_filter_stats_t stats = snoop_filter_->get_coherence_stats();
+        assert(stats.num_writes == expected_num_writes);
+        assert(stats.num_invalidates == expected_num_invalidates);
+        assert(stats.num_writebacks == expected_num_writebacks);
+        addr_t tag = addr >> compute_log2(SNOOPED_FILTER_LINE_SIZE);
+        coherence_table_entry_t *coherence_table_entry = &stats.coherence_table[tag];
+        auto num_sharers = std::count(coherence_table_entry->sharers.begin(),
+                                      coherence_table_entry->sharers.end(), true);
+        assert(num_sharers == expected_num_sharers);
+        assert(block_is_dirty == coherence_table_entry->dirty);
+        snoop_filter_->print_stats();
+    }
+
+private:
+    int num_cores_;
+    snoop_filter_t *snoop_filter_ = new snoop_filter_t;
+    cache_t **snooped_caches_;
+};
+
+void
+unit_test_snoop_filter_two_cores(void)
+{
+    snoop_filter_test_t snoop_filter_test(2);
+    snoop_filter_test.initialize_caches_and_snoop_filter();
+    snoop_filter_test.request_and_check_state(CPU_0, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 0,
+                                              /*block_is_dirty*/ false);
+    snoop_filter_test.request_and_check_state(CPU_1, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 1, false);
+    snoop_filter_test.request_and_check_state(CPU_0, ADDR_A, TRACE_TYPE_WRITE,
+                                              /*num_writes*/ 1,
+                                              /*num_invalidates*/ 1,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 2, false);
```

Comment:
Shouldn't the write invalidate the line in cpu 1 such that there are no sharers and the line is only in cpu 0 in a modified state?  I didn't write this coherence code so I'm less familiar with it: is the sharers vector used in some other way?